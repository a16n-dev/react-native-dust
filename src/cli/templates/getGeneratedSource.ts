import { Project, SourceFile, ts } from 'ts-morph';
import ModuleKind = ts.ModuleKind;
import ScriptTarget = ts.ScriptTarget;

/**
 * This defines the compiler options for generating code at runtime. As a general
 * rule of thumb, this should match up to what is defined in tsconfig.json
 */
const COMPILER_OPTIONS: ts.CompilerOptions = {
  declaration: true,
  module: ModuleKind.ESNext,
  target: ScriptTarget.ES2022,
};

export interface GeneratedFile {
  name: string;
  content: string;
}

export class GeneratedProject {
  private project: Project;

  constructor() {
    this.project = new Project({
      useInMemoryFileSystem: true,
      compilerOptions: COMPILER_OPTIONS,
    });
  }

  /**
   * Adds a new TS file to the project, and returns a reference to the file
   */
  addSourceFile(fileName: string): SourceFile {
    return this.project.createSourceFile(fileName, '', {
      overwrite: true,
    });
  }

  /**
   * Runs diagnostics on the project. If there are any errors in the project,
   * this function will return a string representing the errors
   */
  typeCheck() {
    const diagnostics = this.project.getPreEmitDiagnostics();
    if (diagnostics.length > 0) {
      return this.project.formatDiagnosticsWithColorAndContext(diagnostics);
    }

    return;
  }

  /**
   * Gets all generated files for the entire project
   */
  getGeneratedFiles(): GeneratedFile[] {
    const result = this.project.emitToMemory();
    const files = result.getFiles();

    return files.map((file) => ({
      name: file.filePath,
      content: file.text,
    }));
  }

  /**
   * Gets all of the source (.ts) files for the entire project.
   */
  getSourceFiles(): GeneratedFile[] {
    return this.project.getSourceFiles().map((file) => ({
      name: file.getFilePath(),
      content: file.getFullText(),
    }));
  }
}
