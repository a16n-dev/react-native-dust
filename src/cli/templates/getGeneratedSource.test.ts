import { expect, test } from 'vitest';
import { VariableDeclarationKind } from 'ts-morph';
import { GeneratedProject } from './getGeneratedSource';

test('Source files can be created without error', () => {
  const project = new GeneratedProject();
  const file = project.addSourceFile('example.ts');
  expect(file).toBeDefined();
});

test('generating output of a basic file produces the intended output', () => {
  const project = new GeneratedProject();

  const file = project.addSourceFile('example.ts');

  file.addVariableStatement({
    declarationKind: VariableDeclarationKind.Const,
    isExported: true,
    declarations: [{ name: 'example', initializer: "'Hello world'" }],
  });

  const output = project.getGeneratedFiles();

  // Expect both a single .js and .d.ts file
  expect(output).toHaveLength(2);
  expect(output).toMatchSnapshot();
});
