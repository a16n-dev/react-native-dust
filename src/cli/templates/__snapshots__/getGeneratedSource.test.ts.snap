// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generating output of a basic file produces the intended output 1`] = `
[
  {
    "content": "export const example = 'Hello world';
",
    "name": "/example.js",
  },
  {
    "content": "export declare const example = "Hello world";
",
    "name": "/example.d.ts",
  },
]
`;
